### ISSUE TITLE
Adding mid hook for Recreate deployment strategy fails through oc edit.

### ISSUE THREAD

**Issue by GrahamDumpleton** [openshift/origin#7366]
Create a new application. Edit the deployment configuration using ``oc edit`` and change:

```
  strategy:
    resources: {}
    rollingParams:
      intervalSeconds: 1
      maxSurge: 25%
      maxUnavailable: 25%
      timeoutSeconds: 600
      updatePeriodSeconds: 1
    type: Rolling
```

to:

```
  strategy:
    resources: {}
    recreateParams:
      mid:
        failurePolicy: Abort
        execNewPod:
          command:
            - /bin/sh
          containerName: myapp
    type: Recreate
```

Run ``oc edit`` again and see what was saved and you find:

```
  strategy:
    recreateParams: {}
    resources: {}
    type: Recreate
```

It has not saved the details of the ``mid`` deployment hook.

Now edit the deployment configuration through the UI and once again add the ``mid`` deployment hook. Save change and then edit through the UI once again. The change is still present.

Actually trigger a deployment and the recreate strategy is applied and the ``mid`` deployment hook is run.

Go back and do an ``oc edit`` on the deployment configuration, the ``mid`` deployment hook still isn't shown.

Just for fun, add it back in again via ``oc edit`` and save.

Now go back to the UI again and edit the deployment configuration there, the ``mid`` deployment hook has gone missing from there again.

In summary, adding a ``mid`` deployment hook for ``Recreate`` deployment strategy is not saved when done through ``oc edit``. Further, if change attempted through ``oc edit``, it will wipe any ``mid`` deployment hook that had been able to be setup from the UI.

I have not tried any other types of deployment hooks for ``Recreate`` or ``Rolling``. Is possible affects others, or could be an oversight with ``mid`` deployment hook being new in 1.1.2.

**Comment by kargakis** [comment]
What version of oc are you running? I just tested this with the frontend dc from sample-app that uses the Rolling strategy:
```sh
# actual change
[vagrant@localhost sample-app]$ oc edit dc/frontend
deploymentconfig "frontend" edited
# open editor to verify everything is there
[vagrant@localhost sample-app]$ oc edit dc/frontend
Edit cancelled, no changes made.
# describer
[vagrant@localhost sample-app]$ oc describe dc/frontend
Name:		frontend
Created:	2 minutes ago
Labels:		template=application-template-stibuild
Annotations:	<none>
Latest Version:	1
Triggers:	Image(origin-ruby-sample@latest, auto=true), Config
Strategy:	Recreate
		  Mid-deployment hook (pod type, failure policy: Abort):
		    Container:	myapp
		    Command:	/bin/sh
		    Env:	
Template:
  Selector:	name=frontend
  Replicas:	2
  Containers:
  NAME			IMAGE			ENV
  ruby-helloworld	origin-ruby-sample	ADMIN_PASSWORD=jWnUBGbA,ADMIN_USERNAME=adminHWU,MYSQL_DATABASE=root,MYSQL_PASSWORD=wO5eFvTE,MYSQL_USER=userUV0
Deployment #1 (latest):
	Name:		frontend-1
	Created:	58 seconds ago
	Status:		Running
	Replicas:	0 current / 0 desired
	Selector:	deployment=frontend-1,deploymentconfig=frontend,name=frontend
	Labels:		openshift.io/deployment-config.name=frontend,template=application-template-stibuild
	Pods Status:	0 Running / 0 Waiting / 0 Succeeded / 0 Failed
Events:
  FirstSeen	LastSeen	Count	From				SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  58s		58s		1	{deploymentconfig-controller }			Normal		DeploymentCreated	Created new deployment "frontend-1" for version 1

```

**Comment by GrahamDumpleton** [comment]
It should have been ``oc`` client for 1.1.2 pulled down from github releases for openshift/origin. I do have multiple versions in different directories on my system due to needing to switch between different systems of different versions, so is outside chance that wrong version was used. Let me go back and check, making sure the right version is used.

If it does turn out that it was an older version, then it highlights the issue of people not having the latest, yet no warning is given out by the client that it isn't the correct version.

Does the client at any time issue a warning that you probably should upgrade the client version?

**Comment by GrahamDumpleton** [comment]
Okay. Looks like what happened was that when I was doing testing for #7326 with various ``oc`` versions that I didn't restore things so I was using the latest ``oc`` version with the 1.1.2 backend I was then testing the new mid deployment hooks with. :-(

Would still be nice if ``oc`` client gave some sort of warning about mis match in versions if doesn't already. At least I don't ever recollect seeing such a warning.

**Comment by GrahamDumpleton** [comment]
BTW. Are any details of the hooks and the fact that some are registered meant to appear in the UI, or is all that shows in the UI is the strategy being used?

**Comment by GrahamDumpleton** [comment]
Existing issue about warning on use of old client versions.

* https://bugzilla.redhat.com/show_bug.cgi?id=1306590

**Comment by kargakis** [comment]
> BTW. Are any details of the hooks and the fact that some are registered meant to appear in the UI, or is all that shows in the UI is the strategy being used?

cc: @jhadvig @jwforres

**Comment by kargakis** [comment]
@GrahamDumpleton since there is that bugzilla issue open and the present issue is not really an issue, do you mind closing?

**Comment by jwforres** [comment]
for the UI question, hooks do not show up yet, can be tracked in this story https://trello.com/c/2rDfr6Ma
